ToA lnL first functionas

tschiffer opened the pull request:
Started to inclued the possiblity to use the ToA information for the lnL

magruber committed:
first commit add flags

tschiffer committed:
test

tschiffer committed:
Add flag for ToA cut and lnL

tschiffer committed:
Changed ToA lnL to string input

tschiffer committed:
Reading ToA probabilities return Dataframe and
Energy list

tschiffer committed:
Introduce morphing of ToA lnL data

Vindaar commented on Analysis/ingrid/likelihood.nim:
```diff
@@ -1761,6 +1770,18 @@ proc main(
+#Using this to test the new implementations, need to be removed at some point        
+    echo "path:"
+    ToAProbabilityHists ="../../resources/ToA_P_densitys.csv"
+    var df = newDataFrame()
+    var Energy_list = newSeq[float]()
+    var test = newDataFrame()
+    df = readToAProbabilitys(ToAProbabilityHists, Energy_list)
+    test = getInterpolatedDfToA(df,Energy_list)
```
Just as an explanation: 

You could just as well avoid the `var df / var test` lines and just write `let df = readToA...; let test = getInterp...`.

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -72,6 +72,61 @@ proc getInterpolatedWideDf*(df: DataFrame, num = 1000): DataFrame =
     dfLoc["Variable"] = constantColumn(tup[0][1].toStr, dfLoc.len)
     result.add dfLoc
 
+proc morphToA*(df: DataFrame, lineEnergies: seq, energy: float, offset = 1): (Tensor[int], Tensor[float]) =
```
If you just write `seq` as a type, it is an implicit generic. I.e. the compiler will generate one procedure for each type with which you call this procedure. Say you call it with a `lineEnergies` argument that is `seq[int]` and then later with a `seq[float]`, the compiler generates two procedures.

Better write:

```suggestion
proc morphToA*(df: DataFrame, lineEnergies: seq[float], energy: float, offset = 1): (Tensor[int], Tensor[float]) =
```

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -72,6 +72,61 @@ proc getInterpolatedWideDf*(df: DataFrame, num = 1000): DataFrame =
+  let bins = df["bin", int]
+
+
+  var res = zeros[float](refLowT.size.int)
+  # walk over each bin and compute linear interpolation between
+  let Ediff = abs(lineEnergies[idx] - lineEnergies[idx+offset])
+  for i in 0 ..< bins.size:
+    res[i] = refLowT[i] * (1 - (abs(lineEnergies[idx] - energy)) / Ediff) +
+      refHighT[i] * (1 - (abs(lineEnergies[idx+offset] - energy)) / Ediff)
+  result = (bins, res)
```
Later we can refactor out this type of logic from both the regular and ToA versions of this proc into a `morphImpl` that takes only the reference data, line energies and returns the bins and result. That way we only have the two procedures to do the "data extraction" from the relevant sources and the actual interpolation logic is shared. But I can do that at a later time.

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -72,6 +72,61 @@ proc getInterpolatedWideDf*(df: DataFrame, num = 1000): DataFrame =
+  ## for linear interpolation
+  echo "enter interpolation"
+  let energies = linspace(lineEnergies[0], lineEnergies[^1], num)
+  var test: int
```
`test` is not used.

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -72,6 +72,61 @@ proc getInterpolatedWideDf*(df: DataFrame, num = 1000): DataFrame =
+  echo "enter interpolation"
+  let energies = linspace(lineEnergies[0], lineEnergies[^1], num)
+  var test: int
+  var dfLoc = newDataFrame()
+  var lastBins = zeros[int](0)
+  result = newDataFrame()
+  for idx, E in energies:
+    let (bins, res) = morphToA(df, lineEnergies, E, offset = 1)
+
+    block Sanity:
+      # really the same bins in all Target/Filter combinations? Should be, but check!
+      if lastBins.size > 0:
+        doAssert bins == lastBins
+      lastBins = bins
+    let suffix = "_" & $idx
+    dfLoc["Hist" & $suffix] = res
+
+  dfLoc["Bins"] = lastBins
+  #echo dfLoc  
+  result.add dfLoc
```
Similarly to the above procedure, the actual logic can also be refactored out into a common procedure for regular / ToA cases. Just needs an argument that hands the internal morphing procedure to call. Again, I can do that later.

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -269,6 +324,16 @@ template withXrayRefCuts*(cdlFile, dset: string,
         body
 
+proc readToAProbabilitys*(pathtoToA:string, Energy_list: var seq): DataFrame =
+  let df = readCsv(pathtoToA,sep=' ') 
+  for i in 0 ..< 20000:
+    try:
+      let t: Tensor[float] = df[$i, float]
+      Energy_list.add float(i)
+    except:
+      discard
+  result = df
```
Given that you want the column names of the DF as the energies, better write:

```suggestion
let df = readCsv(pathtoToA,sep=' ') 
let energies = df.getKeys().filterIt(it != "bin").mapIt(it.parseInt)
result = (df, energies)
```

Vindaar commented on Analysis/ingrid/private/likelihood_utils.nim:
```diff
@@ -850,6 +915,12 @@ proc initLikelihoodContext*(
   signalEfficiency: float = 0.0,
+  #ToACut
+  useToACut: bool = false,
+  ToAcutValue: int=0,
```
```suggestion
  ToAcutValue: int = 0,
```

Vindaar commented:
Looking good for your first lines of Nim code! :partying_face: 

Added comments mainly for explanatory purposes.

The next steps would involve:
- modifying the `LikelihoodContext` object to add fields for whether we use ToA cut / ToA LnL / neither and their values here: https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/ingrid_types.nim#L887-L907
- assign these values in the body of `initLikelihoodContext` here: https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/private/likelihood_utils.nim#L883-L963

Once that is done, we can implement the ToA cut (as you proposed) first. Certainly simpler. 

This would involve:
- add the equivalent logic for the ToA cut in this block of code: https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/likelihood.nim#L1309-L1338
  (You might want to add another `isVetoedByToACut` proc or similar that you call there. But for a start, you can just add the logic inline there)
- read the ToA data to have it for the cut in this block a bit above: https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/likelihood.nim#L1260-L1267
  Make sure to only attempt to read the data if we work with a timepix3 detector https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/ingrid_types.nim#L893 and the ToA cut is active! In the other case, just declare an empty `seq[float]`

If this is done and you have added the ToA cut there, we'd need to do 2 other things (but we'll do that once the implementation is there and works):
1. Add the information about how many events were vetoed by the ToA counter here: https://github.com/Vindaar/TimepixAnalysis/blob/master/Analysis/ingrid/likelihood.nim#L127-L140 (needs a counter in the many logic of course). Note that the settings which were used are automatically serialized in that `writeInfos` proc in the `h5f.toH5(...)` call.
2. In downstream code handle the resulting efficiency penalty due to the ToA cut. But for that we first need that number of course (question is whether to estimate it based on a reference dataset or use a fixed number for the beginning). This is so that e.g. in a limit calculation we know the full efficiency.

tschiffer committed:
Changed suggestions from pull request

tschiffer committed:
Added plain ToA cut

Vindaar commented on Analysis/ingrid/likelihood.nim:
```diff
@@ -1265,6 +1279,13 @@ proc filterClustersByVetoes(h5f: var H5File, h5fout: var H5File,
+        discard
+      of Timepix3:
+        if ctx.vetoCfg.useToACut or ctx.vetoCfg.useToAlnLCut:
+          let toaLength = h5f[(chipGroup / "toaLength"), int64].asType(int)
```
```suggestion
          toaLength = h5f[(chipGroup / "toaLength"), int64].asType(int)
```
If you keep the `let`, it shadows the `toaLength` variable defined in the outer scope in the scope of the `case` and `if`. So effectively you never write to the actual `toaLength` variable you care about.

tschiffer committed:
Added couter for num events cut by ToA

tschiffer committed:
minor

tschiffer committed:
remove test commands

Vindaar committed:
[CI] update `mlocate` to `plocate` and webkit version to 4.1

tschiffer committed:
Tested ToA cut. Seems to work

tschiffer committed:
Forgot to add the cut to results

tschiffer committed:
Added morphing with sim data hacky, compiles but
throws out all data maybe refdf need to be log?

tschiffer committed:
minor

tschiffer committed:
files

tschiffer committed:
Changed production of logl cut value calculation.
Added usage of sim data for logl

tschiffer committed:
Started to implement the ToA lnlcut, still some problems

Vindaar committed:
[calib] fix dumb regression I introduced some time ago

I forgot the reason we use ~chipGroups~ is to get the different chips
of a single run. Arguably a better iterator, that takes the run into
account would be better / we could hand the full ~chipBase~ as the
starting path to the ~chipGroups~ iterator. But well, this is cheap.

Vindaar committed:
[tools] add tool to update run type in existing file

Obviously use with care!

Vindaar committed:
[calib] fix regression introduced due to wrong attribute name

The center chip number is just "centerChip"... :)

tschiffer committed:
Finished ToaLnL cut seems to work

tschiffer committed:
bug fix wrong var

tschiffer committed:
Fixed energy from ev to keV

tschiffer committed:
Fixed backward compatibility

Vindaar committed:
minor cleanup of unnecessary code

Vindaar committed:
make Tpx3 reference / sim data paths relative to source location

So that one can run the binary from anywhere on the system and is not
dependent on the relative path to the data files.

Vindaar committed:
fix resource leak due to H5 files not being closed

Vindaar committed:
add TODO about tpx3 timewalk shift

Vindaar committed:
make energy division of ints explicitly float

Vindaar committed:
[calib] catch runs that do not yield any events after gas gain cuts

Vindaar committed:
[logL] update the range of likelihood histogram w/ simulated data

Vindaar committed:
[plots] take out problematic snippet in gas gain over time

Vindaar committed:
[plots] allow setting threshold when switching to discrete in cluster

Vindaar committed:
[plots] improve error message if names.len != files.len & add dataset

... to the rate table

Vindaar committed:
[plots] make gas gain vs time plot usable for non Septemboard
